Sprint 11

Fennec-meow opened the pull request.

Fennec-meow committed:
Constants.

Fennec-meow committed:
sprint_10

Fennec-meow committed:
Some adjustments have been made, as well as the "fetchOAuthToken" function has been changed

Fennec-meow committed:
Corrections:
1. The "OAuth2Service" class
Data transmission has been fixed. Now they are passed not through the "body", but through the "query" parameters.
2. The "SplashViewController" class
The "switchToTabBarController" function has been edited. Changes have been made related to the display of "TabBarViewController" in the main stream.

Fennec-meow committed:
Corrections:
1. Implicit unwrap and other force operations were removed from the project.
2. Constants were described inside the enum Constants.
3. Fixed the "WebView" controller.
4. If there is saved authorization data, the user is sent to the gallery flow, if not— to the authorization flow.
5. After previously successful authorization, the user enters the gallery immediately after "SplashViewController".
6. The completion block is called on the main thread (DispatchQueue.main).
7. A singleton has been added to the "OAuth2Service".
8. Erroneous cases were corrected when receiving optional values such as "URLComponents", "URL", "URLRequest".
9. Correction of the design in accordance with the design of the layout in Figma.

Fennec-meow committed:
Corrections:
1. Implicit unwrap and other force operations were removed from the project.
2. Constants were described inside the enum Constants.
3. Fixed the "WebView" controller.
4. If there is saved authorization data, the user is sent to the gallery flow, if not— to the authorization flow.
5. After previously successful authorization, the user enters the gallery immediately after "SplashViewController".
6. The completion block is called on the main thread (DispatchQueue.main).
7. A singleton has been added to the "OAuth2Service".
8. Erroneous cases were corrected when receiving optional values such as "URLComponents", "URL", "URLRequest".
9. Correction of the design in accordance with the design of the layout in Figma.

Fennec-meow committed:
Commit

Fennec-meow committed:
Sprint_11.
Refactoring

Fennec-meow committed:
no message

Fennec-meow committed:
Changes:
1. Fixed the implementation of "UIBlockingProgressHUD"
2. The radius of the avatar on the profile screen has been changed
3. The transition to the "WebViewController" has been completed

Helvetii951 commented on ImageFeed/Auth/AuthViewController.swift:
```diff
@@ -0,0 +1,115 @@
+//  Created by Kira on 20.01.2025.
+//
+
+import UIKit
+import ProgressHUD
+
+// MARK: - AuthViewControllerDelegate
+
+protocol AuthViewControllerDelegate: AnyObject {
+    func authViewController(_ vc: AuthViewController, didAuthenticateWithCode code: String)
+}
```
В `import ProgressHUD` нет необходимости. Напрямую библиотека не используется. Она обернута в `UIBlockingProgressHUD`.

Helvetii951 commented on ImageFeed/Auth/AuthViewController.swift:
```diff
@@ -0,0 +1,115 @@
+
+// MARK: - AuthViewController
+
+final class AuthViewController: UIViewController {
+    
+    // MARK: UI Components
+    
+    @IBOutlet weak var startButton: UIButton!
+    
+    // MARK: Public Property
+    
+    weak var delegate: AuthViewControllerDelegate?
+    
+    // MARK: Private Property
+    
+    private let showWebViewSegueIdentifier = "ShowWebView"
+    
+    // MARK: Lifecycle
```
Лучше не делать лишние переносы строк, они могут затруднять чтение кода.

Дополнительный материал про кодстайл:  

- [Подробное описание codestyle](https://google.github.io/swift/) от Google
- [От Ray Wanderlich](https://github.com/raywenderlich/swift-style-guide)
- [На русском от команды tutu](https://github.com/tutu-ru/swift-style-guide)

Helvetii951 commented on ImageFeed/Auth/AuthViewController.swift:
```diff
@@ -0,0 +1,115 @@
+        if segue.identifier == showWebViewSegueIdentifier {
+            guard
+                let webViewController = segue.destination as? WebViewController
+            else { fatalError("Failed to prepare for \(showWebViewSegueIdentifier)") }
```
`fatalError` лучше никогда не использовать. Может привести к очень неприятным падениям в будущем. Рекомендую использовать [assertionFailure](https://developer.apple.com/documentation/swift/assertionfaxrilure(_:file:line:)).
Часто в таких случаях пишут ошибки в логи или сервисы статистики и событий.

Helvetii951 commented on ImageFeed/Auth/AuthViewController.swift:
```diff
@@ -0,0 +1,115 @@
+        navigationItem.backBarButtonItem?.tintColor = ColorConstants.ypBlack
+    }
+    
+    @objc func didTapStartButton() {
```
`didTapStartButton` можно сделать приватной. Рекомендую всегда писать функции и свойства с минимальной областью видимости, а затем менять ее, если нужно.

Helvetii951 commented on ImageFeed/Auth/AuthViewController.swift:
```diff
@@ -0,0 +1,115 @@
+    }
+    
+    @objc func didTapStartButton() {
+        // Инициализируем WebViewController
+        let webViewController = WebViewController()
+        webViewController.delegate = self // Если есть делегат, устанавливаем его
+        // Переходим на WebViewController
+        present(webViewController, animated: true)
```
Swift - достаточно выразительный язык. Часто можно понять что происходит и без комментариев.

Но, если тебе в учебе так удобнее, то ОК.

Helvetii951 commented on ImageFeed/Profile/ImegeModel/ProfileImageService.swift:
```diff
@@ -0,0 +1,88 @@
+
+extension ProfileImageService {
+    private func completeFetch(for username: String) -> URLRequest {
+        guard let url = URL(string: "https://api.unsplash.com/users/" + username) else {
```
Очень желательно брать `URL`ы из файла с константами. Так не нужно будет менять код сервиса, если что-то изменится в бэкэнде.

Helvetii951 commented on ImageFeed/Profile/ImegeModel/UserResult.swift:
```diff
@@ -0,0 +1,31 @@
+import Foundation
+
+
+struct UserResult: Codable {
+    let profileImage: ProfileImage
+    
+    enum CodingKeys: String, CodingKey {
+        case profileImage = "profile_image"
+    }
+}
+
+// MARK:  ProfileImage
+
+struct ProfileImage: Codable {
+    let small: String
+    let medium: String
+    let large: String
+    
+    enum CodingKeys: String, CodingKey {
+        case small = "small"
+        case medium = "medium"
+        case large = "large"
+    }
+}
```
Лучше на одну сущность создавать один файл. Так навигация по проекту будет гораздо удобнее. И не нужно будет менять один большой файл из-за небольших изменений в одной из сущностей.

Helvetii951 commented on ImageFeed/Profile/Presentation/ProfileViewController.swift:
```diff
@@ -0,0 +1,251 @@
+//  Created by Kira on 13.01.2025.
+//
+
+import UIKit
+import Foundation
+import Kingfisher
```
`Foundation` уже включен в другие импорты. Можно убрать.

Helvetii951 commented on ImageFeed/Profile/Presentation/ProfileViewController.swift:
```diff
@@ -0,0 +1,251 @@
+    
+    // MARK: Public Property
+    
+    let splashViewController = SplashViewController()
```
Контроллер - тяжеловесный объект со своим жизненным циклом(может реагировать на системные события, кастомные события и пр.). Если есть один экземпляр, который всегда в памяти, можно получить непредсказуемое поведение приложения. Контроллер должен создаваться тогда, когда он нужен. После окончания использования и скрытия он долже выметаться из памяти.

И здесь свойство контроллера не используется. Очень рекомендую убрать.

Helvetii951 commented on ImageFeed/Profile/Presentation/ProfileViewController.swift:
```diff
@@ -0,0 +1,251 @@
+    var username: String? = String()
+    
+    // MARK: Private Property
+    
+    private let swiftKeychainWrapper = SwiftKeychainWrapper()
+    
```
Ненужное свойство. Стоит убрать.

Helvetii951 commented on ImageFeed/Profile/Presentation/ProfileViewController.swift:
```diff
@@ -0,0 +1,251 @@
+    // MARK: Public Property
+    
+    let splashViewController = SplashViewController()
+    
+    var username: String? = String()
+    
```
`String()` = `""`.

Helvetii951 commented on ImageFeed/Profile/Presentation/ProfileViewController.swift:
```diff
@@ -0,0 +1,251 @@
+private extension ProfileViewController {
+    func getAvatar() {
+        UIBlockingProgressHUD.show()
+        profileImageService.fetchProfileImageURL(username ?? String()) { [weak self] result in
```
Нет смысла посылать запрос с пустым именем. Он все-равно завершится в ошибкой. Лучше просто прервать выполнение функции, если имени нет.

Helvetii951 commented on ImageFeed/Profile/Presentation/ProfileViewController.swift:
```diff
@@ -0,0 +1,251 @@
+    
+    // MARK: The exit button from your personal account
+    
+    @objc func didTapLogoutButton() {
```
Можно сделать `private`.

Helvetii951 commented on ImageFeed/SplashViewController/SplashViewController.swift:
```diff
@@ -0,0 +1,231 @@
+                        }
+                        
+                    }
+                    UIBlockingProgressHUD.dismiss()
+                    self.switchToTabBarController()
+                case .failure:
+                    UIBlockingProgressHUD.dismiss()
```
`UIBlockingProgressHUD.dismiss()` уже есть до `switch`. Здесь дублирование чуть избыточно.

Helvetii951 commented on ImageFeed/SplashViewController/SplashViewController.swift:
```diff
@@ -0,0 +1,231 @@
+    
+    // MARK: UI Components
+    
+    private var vectorImageView: UIImageView = {
+        let imageView = UIImageView(image: UIImage(named: "Vector"))
+        imageView.contentMode = .scaleAspectFit
+        imageView.translatesAutoresizingMaskIntoConstraints = false
+        return imageView
+    }()
```
Рекомендую всегда использовать [ленивые свойства](https://www.swiftbysundell.com/articles/using-lazy-properties-in-swift/)для объявления подобного  В нашем случае это какого-то буста не даст. Но, если какая-то вьюха/свойство не будет использоваться, можно сэкономить память.
Подход с ленивыми свойствами используют во многих хороших компаниях.

Fennec-meow committed:
Changes:
1. Changes have been made to the "AuthViewController":
a. Unused "import ProgressHUD" has been removed;
b. The "prepare(for segue: UIStoryboardSegue, sender: Any?" method" changes were made during an unsuccessful transition;
c. In the "func configureNavigationBar()" method, "private" has been removed.
2. Changes have been made to the "SplashViewController":
a. The "lazy" prefix has been added to the "vectorImageView" component;
b. In the "fetchOAuthToken" method, if a token is received incorrectly, the "case.failure:" has been changed to "case .failure(let error):", slightly changed the print and commented out the call to "self.showAlert()"
